src/test/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapperTest.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.bridge.mqtt.core;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.strimzi.kafka.bridge.mqtt.utils.MappingRule;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Unit tests for {@link MqttKafkaMapper}
+ */
+public class MqttKafkaMapperTest {
+
+    private ArrayList<MappingRule> mappingRules;
+    private MqttKafkaMapper mqttKafkaMapper;
+    //define the mapping rules as a JSON string. Add more rules if needed.
+    private static final String TOPIC_MAPPING_RULES_JSON =
+            "[" +
+                    "{\"kafkaTopic\":\"building_{building}_room_{room}\",\"mqttTopic\":\"building/{building}/room/{room}/#\"}," +
+                    "{\"kafkaTopic\":\"sensor_data\",\"mqttTopic\":\"sensors/+/data\"}," +
+                    "{\"kafkaTopic\":\"devices_{device}_data\",\"mqttTopic\":\"devices/{device}/data\"}," +
+                    "{\"kafkaTopic\":\"fleet_{vehicle}\",\"mqttTopic\":\"fleet/{fleet}/vehicle/{vehicle}/#\"}," +
+                    "{\"kafkaTopic\":\"building_{building}_others\",\"mqttTopic\":\"building/{building}/#\"}," +
+                    "{\"kafkaTopic\":\"sensor_others\",\"mqttTopic\":\"sensors/#\"}," +
+                    "{\"kafkaTopic\":\"building_others\",\"mqttTopic\":\"building/#\"}" + "]";
+
+    /**
+     * Load all the mapping rules before running the tests.
+     */
+    @Before
+    public void setUp() {
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            mappingRules = mapper.readValue(TOPIC_MAPPING_RULES_JSON, new TypeReference<>() {
+            });
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        mappingRules.sort(Comparator.comparing(MappingRule::getMqttTopicPatternLevels).reversed());
```

Comment:
Clever :)

Reply:
I don't think this is enough to have the MQTT pattern with `+` taking precedence on the ones with `#` at the end.
If I got it right, this comparator is based on counting how many `/` are in the pattern (so what you call "levels") but ... take these two MQTT patterns as an example:

`/foo/+/bar/+` and `/foo/+/bar/#`

They have the same "levels" (same number of  `/` character) but ...

If the producer is sending to `/foo/10/bar/5` we want it being matched with `/foo/+/bar/+` and not `/foo/+/bar/#`. How we are sure that with the comparator you put in place we have this order?